docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+users are likely to create topics for occasional or one-off use.
+
+# Motivation
+
+## Why are we doing this?
+
+Partitions have some runtime resource overhead, even if they aren't being
+used, and haven't been used for some time.
+
+To support public-facing Kafka services on which many users might create
+and then abandon idle topics, we could implement a "hibernate" mode
+for partitions that haven't been used for a long time.
+
+## What use cases does it support?
+
+Implementing a low-cost/no-cost public service, in which developers are likely
+to create unused topics with which to experiment.
+
+## What is the expected outcome?
+
+Systems will run smoothly with a higher number of partitions than
+they could before, as long as some of those partitions are idle.
+
+# Guide-level explanation
+
+## Introducing new named concepts.
+
+`Hibernated topic`: a topic whose partitions do not have active raft
+groups running in the cluster.  It cannot serve user I/O (aside from
+metadata requests) until it is woken up.
+
+`Waking a topic`: a control plane operation to change the status of
+the topic, followed by raft elections for all partitions within the
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
+on the system, is driven by controller log commands.  New commands
+are added for hibernate and wake.  Nodes in a partition's raft group
+notice the hibernation of the topic, and shut down the raft group.
+
+There is a "hibernating" intermediate state where members of the
+partitions' raft groups are responsible for refusing writes, and
+reporting the highest offset back to the controller so that it can
+reliably store the frozen offset state of the topic when storing
+the controller message that finalizes the hibernation.
+
+Once the topic is hibernated, all its partitions will report a
+fictional leader, so that clients know where to send metadata requests,
+and so that monitoring systems do not incorrectly flag the partition
+as being in a problematic state.
+
+On startup, a hibernated partition's storage resources are not
+recovered or replayed.  Hibernated partitions are not subject to
+continuous retention policies, BUT the controller can notice if
+a partition has been hibernated for longer than its retention.ms
+limit, and wake it up for long enough to do housekeeping.
+
+We should still be able to serve metadata requests for idle topics,
+because otherwise any monitoring system will end up re-awakening
+the topics whenever querying them.
+
+## How do we decide when to hibernate?
+
```

Comment:
Alternative: don't decide to hibernate, aim to demand-load resources and "hibernate resources LRU behind the scenes" as needed to stay within resource limits.

Reply:
Yep, that's the "Radically improved resource management elsewhere" section in Alternatives.